pkg/metaclient/kvclient/etcdkv/etcd_client_test.go

```diff
@@ -0,0 +1,310 @@
+}
+
+func clearKeySpace(ctx context.Context, cli metaclient.KVClient) {
+	cli.Delete(ctx, "", metaclient.WithFromKey())
+}
+
+func (suite *SuiteTestEtcd) TestBasicKV() {
+	conf := &metaclient.Config{
+		Endpoints: []string{suite.endpoints},
+	}
+	t := suite.T()
+	cli, err := NewEtcdImpl(conf)
+	require.Nil(t, err)
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	// get a non-existed key
+	grsp, gerr := cli.Get(ctx, "hello")
+	require.Nil(t, gerr)
+	cluster := grsp.Header.ClusterID
+	require.NotEmpty(t, cluster)
+	require.Len(t, grsp.Kvs, 0)
+
+	// put a key and get it back
+	prsp, perr := cli.Put(ctx, "hello", "world")
+	require.Nil(t, perr)
+	require.Equal(t, cluster, prsp.Header.ClusterID)
+	grsp, gerr = cli.Get(ctx, "hello")
+	require.Nil(t, gerr)
+	require.Len(t, grsp.Kvs, 1)
+	require.Equal(t, "world", string(grsp.Kvs[0].Value))
+
+	// delete a key, get it back, put the same key and get it back
+	drsp, derr := cli.Delete(ctx, "hello")
+	require.Nil(t, derr)
+	require.Equal(t, cluster, drsp.Header.ClusterID)
+	grsp, gerr = cli.Get(ctx, "hello")
+	require.Nil(t, gerr)
+	require.Len(t, grsp.Kvs, 0)
+	prsp, perr = cli.Put(ctx, "hello", "world again")
+	require.Nil(t, perr)
+	require.Equal(t, cluster, prsp.Header.ClusterID)
+	grsp, gerr = cli.Get(ctx, "hello")
+	require.Nil(t, gerr)
+	require.Len(t, grsp.Kvs, 1)
+	require.Equal(t, "world again", string(grsp.Kvs[0].Value))
+	clearKeySpace(ctx, cli)
+	cli.Close()
+	// [TODO] using failpoint to test basic kv timeout and cancel
+}
+
+func prepareDataForRangeTest(ctx context.Context, t *testing.T, cli metaclient.KVClient) {
+	prsp, perr := cli.Put(ctx, "hello1", "world1")
+	require.Nil(t, perr)
+	cluster := prsp.Header.ClusterID
+	require.NotEmpty(t, cluster)
+	prsp, perr = cli.Put(ctx, "hello2", "world2")
+	require.Nil(t, perr)
+	prsp, perr = cli.Put(ctx, "interesting", "world")
+	require.Nil(t, perr)
+	prsp, perr = cli.Put(ctx, "dataflow", "engine")
+	require.Nil(t, perr)
+	prsp, perr = cli.Put(ctx, "TiDB", "component")
+	require.Nil(t, perr)
+}
+
+func (suite *SuiteTestEtcd) TestKeyRangeOption() {
+	conf := &metaclient.Config{
+		Endpoints: []string{suite.endpoints},
+	}
+	t := suite.T()
+	cli, err := NewEtcdImpl(conf)
+	require.Nil(t, err)
+	defer cli.Close()
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	// test get key range(WithRange/WithPrefix/WithFromKey)
+	prepareDataForRangeTest(ctx, t, cli)
+	defer clearKeySpace(ctx, cli)
+
+	grsp, gerr := cli.Get(ctx, "hello", metaclient.WithRange("s"))
+	require.Nil(t, gerr)
+	require.Len(t, grsp.Kvs, 3) // hello1, hello2, interesting
+	require.Equal(t, string(grsp.Kvs[0].Key), "hello1")
+	require.Equal(t, string(grsp.Kvs[0].Value), "world1")
+	require.Equal(t, string(grsp.Kvs[1].Key), "hello2")
+	require.Equal(t, string(grsp.Kvs[1].Value), "world2")
+	require.Equal(t, string(grsp.Kvs[2].Key), "interesting")
+	require.Equal(t, string(grsp.Kvs[2].Value), "world")
```

Comment:
Better to make the test case into table format, such as
```go
type kv struct {
    key string
    value string
}

input = []struct{
    kv kv
} {
    {"key1", "value1"},
    {"key2", "value2"},
}

query := []struct {
    key string    
    opts Op
    expected []kv 
} {
    ....
}
```
The case will be more clear and easy to maintain.